Training24/NumConversion.c

```diff
@@ -0,0 +1,141 @@
+   return (*num == '\0');
+}
+
+/// <summary>Function to convert an integer into hexadecimal representation.</summary>
+void BinaryToHex (const char binary[], char hex[], int bitWidth) {
+   int hexDigits = bitWidth / BITS_PER_HEX_DIGIT;
+   int binaryIndex = 0;
+   for (int i = 0; i < hexDigits; i++) {
+      int n = 0;
+      for (int j = 0; j < BITS_PER_HEX_DIGIT; j++) n = (n << 1) | (binary[binaryIndex++] - '0');
+      hex[i] = (n < 10) ? '0' + n : 'A' + (n - 10);
+   }
+   hex[hexDigits] = '\0';  // Null-terminate the string
+}
+
+/// <summary>Function to determine the appropriate bit width based on the integer value.</summary>
+int DetermineBitWidth (int number) {   
+   if (number == INT_MIN) return 32;   // Determine the minimum bit width required to represent the number
+   if (number < -(1 << 15) || number >= (1 << 15)) return 32;
+   if (number < -(1 << 7) || number >= (1 << 7)) return 16;
+   return 8;   // Default bit width
+}
+
+/// <summary> Test binary and hexadecimal conversions with various inputs and checks for correctness.</summary>
+void TestCases () {
+   struct {
+      long long number;
+      char expectedBinary[MAX_BINARY_DIGITS]; char expectedHex[MAX_HEX_DIGITS + 1];
+   } testCases[] = {
+       {0, "00000000", "00"},
+       {-128, "10000000", "80"},
+       {255, "0000000011111111", "00FF"},
+       {INT_MAX, "01111111111111111111111111111111", "7FFFFFFF"},
+       {INT_MIN, "10000000000000000000000000000000", "80000000"},
+       {2147483649ll, "00000000000000000000000000000000", "00000000"},   
+       {-2147483649ll, "10000000000000000000000000000000", "80000000"}   // Error cases: Out of range
+   };
+   int numTests = sizeof (testCases) / sizeof (testCases[0]);  
+   for (int i = 0; i < numTests; i++) {
+      long long number = testCases[i].number;
+      if (IsWithinRange (number)) {
+         int numberRange = (int)number;
+         int bitWidth = DetermineBitWidth (numberRange);
+         char binary[MAX_BINARY_DIGITS]; char hex[MAX_HEX_DIGITS + 1];
+         ConvertToBinary (numberRange, bitWidth, binary);
+         BinaryToHex (binary, hex, bitWidth);
+         printf ("Number = %lld\n", number);
+         printf ("Binary value: %s\n", binary);
+         printf ("Binary Test: %s\n", strcmp (binary, testCases[i].expectedBinary) == 0 ? GREEN"PASS" RESET : RED "FAIL" RESET);
+         printf ("Hexadecimal value: %s\n", hex);
+         printf ("Hexadecimal Test: %s\n\n", strcmp (hex, testCases[i].expectedHex) == 0 ? GREEN "PASS" RESET : RED "FAIL" RESET);
+      }
+      else {
+         printf ("Number = %lld\n", number);
+         printf (YELLOW "Error: Number is out of 32-bit range.\n" RESET "\n");
+      }
+   }
+}
+int main () {
+   long long number;
+   char binary[MAX_BINARY_DIGITS];
+   char hex[MAX_HEX_DIGITS + 1];
+   TestCases ();
+   while (true) {
+      printf ("Enter an integer: ");
+      if (GetValidInteger (&number)) {
+         if (IsWithinRange (number)) {

```

Comment:
```suggestion
         if (!IsWithinRange (number)) printf (YELLOW "Error: Number is out of 32-bit range.\n" RESET "\n");

```
See [Guard](https://en.wikipedia.org/wiki/Guard_(computer_science)).